src/PrefixTree.java

```diff
@@ -30,55 +27,51 @@ public boolean contains(CharSequence word){
     }
 
     public boolean add(CharSequence word){
-        Helper helper = traverse(word);
-        Node previousNode = helper.node;
+        TraverseResult helper = traverse(word);
+        Node currentNode = helper.node;
         if(helper.index == -1){
             System.out.println("Word has been added already. " + word);
             return false;
         }
-        else{
-            for (int i = helper.index; i < word.length(); i++) {
-                if(word.length() == 1){
-                    break;
-                }
-                Edge edge = new Edge();
-                Node node = new Node();
-                node.setValue(word.charAt(i));
-                edge.setReference(node);
-                previousNode.addEdge(edge);
-                previousNode = node;
-            }
-            previousNode.setFinal(true);
+        if(word.length() == 1){
+            currentNode.setFinal(true);
             return true;
         }
+        for (int i = helper.index; i < word.length(); i++) {
+            Node nextNode = new Node(word.charAt(i));
+            Edge edge = new Edge(nextNode);
+            currentNode.addEdge(edge);
+            currentNode = nextNode;
+            }
+            currentNode.setFinal(true);
+            return true;
     }
 
-    private Helper traverse(CharSequence word){
-        Helper helper = new Helper();
-        Node previousNode = root;
-        for (int i = 0; i < word.length(); i++) {
-            boolean hasEdge = false;
-            ArrayList<Edge> nodeEdges = previousNode.getEdges();
-            for (Edge nodeEdge : nodeEdges) {
-                if(nodeEdge.getReferences().getValue() == word.charAt(i)){
-                    previousNode = nodeEdge.getReferences();
-                    hasEdge = true;
-                    break;
-                }
-            }
-            if(!hasEdge){
-                helper.index = i;
-                helper.node = previousNode;
-                return helper;
+    private TraverseResult traverse(CharSequence word){
+        TraverseResult traverseResult = new TraverseResult();
+        traverseResult.node = root;
+        for (traverseResult.index = 0; traverseResult.index < word.length(); traverseResult.index++) {
+            AtomicBoolean hasEdge = new AtomicBoolean(false);
+            ArrayList<Edge> nodeEdges = traverseResult.node.getEdges();
+
+            nodeEdges.forEach(
+                    edge -> {
+                        if (edge.getNode().getValue() == word.charAt(traverseResult.index)) {
+                            traverseResult.node = edge.getNode();
+                            hasEdge.set(true);
+                        }
+                    }
+            );
+
+            if(!hasEdge.get()){
+                return traverseResult;
             }
```

Comment:
Previous version is much better, I think. The lambda only complicates things with this way of using it. Lambdas in Java are glimmers of functional programming applied to the imperative way of the language. Mixing the two approaches is rarely a good idea.

Dropping `hasEdge` will be a great improvement here.
We are essentialy looking for a particular edge. See [here](https://stackoverflow.com/questions/22940416/fetch-first-element-of-stream-matching-the-criteria) for a clue.